### PR TITLE
feat: finish MLIR scaffolding for EXLA.Defn

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: exla/lib/exla/computation.ex
Comment (polvalente) [inline exla/lib/exla/computation.ex:85]: @josevalim we need to think about how to deal with this. EXLA wraps the result in a tuple automatically, so I just hardcoded this output value.

Comment (polvalente) [inline exla/lib/exla/computation.ex:85]:
```suggestion
      # TO-DO (mlir): do not hardcode this single-item tuple output type
      EXLA.Shape.make_tuple_shape([ret_shape])
```

File: exla/lib/exla/builder.ex
Comment (polvalente) [inline exla/lib/exla/builder.ex:74]:
```suggestion
```

File: exla/lib/exla/builder.ex
Comment (polvalente) [inline exla/lib/exla/builder.ex:70]:
```suggestion
      # TO-DO: do not hardcode fetching just the first item as the output
      EXLA.MLIR.Value.get_tuple_element(val, 0)
```

File: exla/lib/exla/builder.ex
Comment (polvalente) [inline exla/lib/exla/builder.ex:71]: we probably want to deal with this in a different manner

File: exla/lib/exla/defn.ex
Comment (polvalente) [inline exla/lib/exla/defn.ex:851]: In theory, we just need to add the MLIR NIFs for all the `@bin_op`s and they'll be ready :)

Comment (polvalente) [reply exla/lib/exla/defn.ex:851]: We should leave it to after this and sm-metal are merged to main

File: exla/lib/exla/computation.ex
Comment (josevalim) [reply exla/lib/exla/computation.ex:86]: Is there an issue in always making it a tuple? A single-element tuple has no overhead afaik and it helped address bugs between CPU and GPU.

File: exla/lib/exla/builder.ex
Comment (josevalim) [inline exla/lib/exla/builder.ex:55]: I would hide this version of `new`.

File: exla/lib/exla/builder.ex
Comment (josevalim) [inline exla/lib/exla/builder.ex:60]: Btw, this is used in nested scopes such as `if`, `cond`. So I imagine we will add another branch in the future. :)

File: exla/lib/exla/computation.ex
Comment (josevalim) [inline exla/lib/exla/computation.ex:76]: Can we just call it `return_shape`?

File: exla/lib/exla/builder.ex
Comment (josevalim) [inline exla/lib/exla/builder.ex:39]: I would move the creation of Type.new fully to the C code right now. Have the C function accept EXLA.Shape and convert to Type there, get rid of the Type module. Otherwise it will be confusing keeping both `mlir_ret_type` and `xla_ret_shape`. We just need one type and the easier for now is EXLA.Shape. We can revisit later.

File: exla/lib/exla/computation.ex
Comment (polvalente) [reply exla/lib/exla/computation.ex:76]: now that we're removing MLIR-specific things like MLIR.Type, yes

File: exla/lib/exla/builder.ex
Comment (polvalente) [reply exla/lib/exla/builder.ex:60]: Yes, I only implemented the minimal changes to get things to work. The new branch would have been untested, so I didn't bother

File: exla/lib/exla/builder.ex
Comment (polvalente) [reply exla/lib/exla/builder.ex:55]: hide === defp it?

File: exla/lib/exla/computation.ex
Comment (polvalente) [reply exla/lib/exla/computation.ex:86]: Not exactly an issue. The problem right now is that ret_shape was taken from a tensor-return. I assume that returning a tuple or another container would need something else for actually handling here.

File: exla/lib/exla/builder.ex
Comment (polvalente) [reply exla/lib/exla/builder.ex:55]: I turned it into a defp. Lots of tests needed changing, but I think that's expected.

File: exla/lib/exla/computation.ex
Comment (polvalente) [reply exla/lib/exla/computation.ex:76]: I removed MLIR.Type, and this is now just return_shape :)

File: exla/lib/exla/builder.ex
Comment (polvalente) [reply exla/lib/exla/builder.ex:39]: Done!

**Review by josevalim — APPROVED** [review]

**Pull request merged**
